ttylog.8
.TH TTYLOG 8 "2016-06-02" "" ""
.SH NAME
ttylog \- serial device logger
.SH SYNOPSIS
.B ttylog
[-b|--baud] [-d|--device] [-f|--flush] [-t|--timeout] > /path/to/logfile
.PP
If you are not using the timeout option, you can stop it running by pressing a
ctrl-c when it's going to the screen or doing "kill -HUP nnnn" (where nnnn is
the process ID for ttylog) if running in the background.
.SH DESCRIPTION
This program writes everything that comes from a serial device like /dev/ttyS1
onto stdout. You can specify the device and the baud rate of the device.
.SH OPTIONS
ttylog follows the GNU command line syntax, with long options
starting with two dashes('-').
A summary of the options is included below:
.TP
.B -h, --help
Displays a little help
.TP
.B -v, --version
Version number of ttylog plus Copyright information
.TP
.B -b, --baud
Baud rate of the device. Available baud rates:
300, 1200, 2400, 4800, 9600, 19200, 38400, 57600 and 115200
.TP
.B -d --device
The serial device. For example /dev/ttyS1
.TP
.B -f --flush
Output buffers are flushed after every write.
.TP
.B -t --timeout
How long to run ttylog, in seconds.
.SH AUTHOR
This manual page was originally written by Tibor Koleszar <t.koleszar@somogy.hu>,
for the Debian GNU/Linux system.  Modifications and updates written by
Robert James Clay <jame@rocasa.us>.


ChangeLog
2016-6-04  Robert James Clay <jame@rocasa.us>

    * README.md: Minor edits to content of file.


2016-6-03  Robert James Clay <jame@rocasa.us>  0.28

    * Release as ttylog version v0.28.


2016-6-02  Robert James Clay <jame@rocasa.us>

    * README.md: Add Usage, Code, Support, and other information.
    * ttylog.8: Add information to the SYNOPSIS section.


2016-6-01  Robert James Clay <jame@rocasa.us>

    * ttylog.c: ttylog.8: Rewrite description of the '-t|--timeout' option.
    * ttylog.c: Remove the '-r|--raw' command line option, at least for now.
    * ttylog.8: Add SYNOPSIS and edit the DESCRIPTION & OPTIONS sections.


2016-5-30  Robert James Clay <jame@rocasa.us>

    * ttylog.c: Include the changes for ttylog.c from the patch associated with
      Debian bug # 739733, thus closing the bug. The changes were intended to:
      - Fix extra newlines being added to output.
      - Fix dropping of lines at high baud rates.
      - Fix dropping of first input line.
      Thanks to Doug Brunner <dbrunner@dbus.com>.
    * ttylog.c: Update copyright infomation presented by '--version' option.
    * COPYRIGHT: Update copyright years for jame@rocasa.us and add information
      for alexander.fust.info@gmail.com.


2016-5-25  Alexander (MrMontag) Fust <alexander.fust.info@gmail.com>  0.27

    * CMakeList.txt: Finalize and activate the CPack directives 


2016-5-13  Robert James Clay <jame@rocasa.us>

    * ttylog.c: Remove extraneous character at line 130.
    * config.h.in: Remove extraneous character at line 3.
    * ttylog.c: Update the copyright lines for the current authors.


2016-4-17  Alexander (MrMontag) Fust <alexander.fust.info@gmail.com>

    * ttylog.c: bug-fixes and improvements for the new '-t' & '-r' options.
    * CMakeList.txt: CPack directives deactivated by commenting them out.


2016-4-09  Alexander (MrMontag) Fust <alexander.fust.info@gmail.com>

    * CMakeList.txt, config.h.in, ttylog.c: Improve the CMake configuration.


2016-4-05  Alexander (MrMontag) Fust <alexander.fust.info@gmail.com>

    * ttylog.c, ttylog.8: New 'raw' option added, as '-r' or '--raw'.
    * ttylog.c, ttylog.8: Change the '-t' long option from '--time  Seconds to run'
      to '--timeout  Time to run'.


2016-4-04  Alexander (MrMontag) Fust <alexander.fust.info@gmail.com>

    * ttylog.8, ttylog.c: New 'time' option added, as '-t' or '--time'.
    * CMakeList.txt: Change the setting of C_FLAGS and add CPACK related directives.


2015-6-03  Robert James Clay <jame@rocasa.us>  0.26

    * ttylog.c: Correct truncating of modem_device string, implementing ttylog
      patch #3 (thanks to Daniel Flinkmann). Closes Debian Bug #731086.
    * ttylog.c: Rewrite the "-h|--help" option code, including the expansion
      of the "Usage:" line.
    * ttylog.c: Rewrite the '-v|--version' option code, including the addition
      of the License and Copyright information.
    * ChangeLog: Correct bad date entry for v0.1.b-3.


2014-7-29  Robert James Clay <jame@rocasa.us>

    * ttylog.8: Rewrite the '-f --flush' option description in ttylog.8.


2013-6-19  Robert James Clay <jame@rocasa.us>

    * README.md: Add initial version of the README.md file.


2013-2-14  Robert James Clay <jame@rocasa.us>  0.25

    * ttylog.c: Corrections to the code adding a NULL to modem_device variable.

    * CMakeLists.txt: Change CMAKE_C_FLAGS default settings to be Debug build
      settings by adding them to CMAKE_C_FLAGS_DEBUG instead.


2013-2-14  Robert James Clay <jame@rocasa.us>  0.24

    * ttylog.c: Cleanup of tabs and whitespace.

    * ttylog.c: Ensure that the modem_device variable is NULL terminated before
      being used.


2013-2-14  Robert James Clay <jame@rocasa.us>  0.23

    * Correct how CMAKE_C_FLAGS is getting set in CMakeLists.txt.


2013-2-13  Robert James Clay <jame@rocasa.us>  0.22

    * Change the ttylog build system to use CMake.
    * Add an INSTALL file with information about the CMake based build system.
      Include the legacy Makefile as a separate section in the file.


2013-1-29  Robert James Clay <jame@rocasa.us>  0.21

    * Do not explicitly define the CC variable in the Makefile.
    * Add the CPPFLAGS variable to object build line in the Makefile.
    * Change CFLAGS to LDFLAGS in the executable build line in Makefile.


2013-1-29  Robert James Clay <jame@rocasa.us>  0.20.0

    * Change the version numbering scheme to 'm.n.p'; where "m" is the major
      version number, "n" is the minor version number, and "p" is a patch
      version number. Set this version as v0.20.0, derived by mapping the old
      versions to new minor versions starting from 0.1.0 and counting bug
      fixes as separate minor versions. For details see ttylog bug # 5, which
      this closes.


2013-1-21  Robert James Clay <jame@rocasa.us>  0.1.f

    * Use standard variable CFLAGS instead of C_ARGS in the Makefile and change
      how it is initially set there.
    * Fix issue where ttylog did not test length of device. Bug # 553945 reported
      in Debian regarding an issue where ttylog does not check the length of the
      device name, which caused a segfault if the name was too long. Reference also
      ttylog bug # 3.


2012-09-25  Robert James Clay <jame@rocasa.us>  0.1.e

    * Add missing '-f' option information to ttylog man page.


2011-02-16  Robert James Clay <jame@rocasa.us>  0.1.d

    * Corrected a misspelling of the word 'version' in ttylog.c.
    * Change to using jame@rocasa.us for my email address.
    * Corrected the FSF mailing address in the COPYRIGHT file.
    * Explicitly noted copyright information at beginning of ttylog.c.
    * Derived this ttylog ChangeLog from debian/changelog, which is where
      all changes were being noted previously.


2008-06-26  Robert James Clay <jame@rocasa.org>  0.1.c

    * Added support for 4800 baud (Closes: Debian bug#203356)
    * Remove unneeded printf for flushed option. (Closes: Debian bug#267872)

ttylog.c
/* ttylog - serial port logger
 Copyright (C) 1999-2002  Tibor Koleszar <oldw@debian.org>
 Copyright (C) 2008-2016  Robert James Clay <jame@rocasa.us>
 Copyright (C)      2016  Alexander (MrMontag) Fust <MrMontagOpenDev@gmail.com>

 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/
#include <sys/stat.h>
#include <termios.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <fcntl.h>
#include <signal.h>
#include <time.h>
#include <errno.h>

#include "config.h"

#define BAUDN 9

char flush = 0;

char *BAUD_T[] =
{"300", "1200", "2400", "4800", "9600", "19200", "38400", "57600", "115200"};

int BAUD_B[] =
{B300, B1200, B2400, B4800, B9600, B19200, B38400, B57600, B115200};

int
main (int argc, char *argv[])
{
  FILE *logfile;
  fd_set rfds;
  int retval, i, j, baud = -1;
  timer_t timerid;
  struct sigevent sevp;
  sevp.sigev_notify = SIGEV_SIGNAL;
  sevp.sigev_signo = SIGINT;
  int fd;
  char line[1024], modem_device[512];
  struct termios oldtio, newtio;

  modem_device[0] = 0;

  if (argc < 2)
    {
      printf ("%s: no params. try %s -h\n", argv[0], argv[0]);
      exit (0);
    }

  for (i = 1; i < argc; i++)
    {

      if (!strcmp (argv[i], "-h") || !strcmp (argv[i], "--help"))
        {
          printf ("ttylog version %s\n", TTYLOG_VERSION);
          printf ("Usage:  ttylog [-b|--baud] [-d|--device] [-f|--flush] [-t|--timeout] > /path/to/logfile\n");
          printf (" -h, --help	This help\n -v, --version	Version number\n -b, --baud	Baud rate\n");
          printf (" -d, --device	Serial device (eg. /dev/ttyS1)\n -f, --flush	Flush output\n");
          printf (" -t, --timeout  How long to run, in seconds.\n");
          printf ("ttylog home page: <http://ttylog.sourceforge.net/>\n\n");
          exit (0);
        }

      if (!strcmp (argv[i], "-v") || !strcmp (argv[i], "--version"))
        {
          printf ("ttylog version %s\n", TTYLOG_VERSION);
          printf ("Copyright (C) 2016 Robert James Clay <jame@rocasa.us>\n");
          printf ("Copyright (C) 2016 Alexander (MrMontag) Fust <alexander.fust.info@gmail.com>\n");
          printf ("Copyright (C) 2002 Tibor Koleszar <oldw@debian.org>\n");
          printf ("License GPLv2+: <http://www.gnu.org/licenses/old-licenses/gpl-2.0.html>\n");
          printf ("This is free software: you are free to change and redistribute it.\n");
          printf ("There is NO WARRANTY, to the extent permitted by law.\n");
          exit (0);
        }

      if (!strcmp (argv[i], "-f") || !strcmp (argv[i], "--flush"))
        {
          flush = 1;
        }

      if (!strcmp (argv[i], "-b") || !strcmp (argv[i], "--baud"))
        {
          if (argv[i + 1] != NULL)
            {
              for (j = 0; j < BAUDN; j++)
                if (!strcmp (argv[i + 1], BAUD_T[j]))
                  baud = j;
            }
          if (baud == -1)
            {
              printf ("%s: invalid baud rate %s\n", argv[0], argv[i + 1]);
              exit (0);
            }
        }

    if (!strcmp (argv[i], "-d") || !strcmp (argv[i], "--device"))
      {
        if (argv[i + 1] != NULL)
          {
            memset (modem_device, '\0', sizeof(modem_device));
            strncpy (modem_device, argv[i + 1], sizeof(modem_device)-1);
            modem_device[(sizeof(modem_device)-1)] = '\0';
          }
        else
          {
          }
      }

    if (!strcmp (argv[i], "-t") || !strcmp (argv[i], "--timeout"))
      {
        if (argv[i + 1] == NULL)
          {
            printf ("%s: invalid time span %s\n", argv[0], argv[i + 1]);
            exit(0);
          }
        if (timer_create (CLOCK_REALTIME, &sevp, &timerid) == -1)
          {
            printf ("%s: unable to create timer: %s\n", argv[0], strerror(errno));
            exit (0);
          }
        struct itimerspec new_value;
        new_value.it_interval.tv_sec = 0;
        new_value.it_interval.tv_nsec = 0;
        int sec = atoi(argv[i + 1]);
        if (!sec)
          {
            printf ("%s: invalid time span %s\n", argv[0], argv[i + 1]);
            exit(0);
          }
        new_value.it_value.tv_sec = atoi(argv[i + 1]);
        new_value.it_value.tv_nsec = 0;
        if (timer_settime(timerid, 0, &new_value, NULL) == -1)
          {
            printf ("%s: unable to set timer time: %s\n", argv[0], strerror(errno));
            exit (0);
          }
      }
    }

  if (!strlen(modem_device)) {
    printf ("%s: no device is set. Use %s -h for more information.\n", argv[0], argv[0]);
    exit (0);
  }

  logfile = fopen (modem_device, "rb");
  if (logfile == NULL)
    {
      printf ("%s: invalid device %s\n", argv[0], modem_device);
      exit (0);
    }
  fd = fileno (logfile);

  tcgetattr (fd, &oldtio);	/* save current serial port settings */
  bzero (&newtio, sizeof (newtio));	/* clear struct for new port settings */

  newtio.c_cflag = BAUD_B[baud] | CRTSCTS | CS8 | CLOCAL | CREAD;
  newtio.c_iflag = IGNPAR | IGNCR;
  newtio.c_oflag = 0;
  newtio.c_lflag = ICANON;

  tcflush (fd, TCIFLUSH);
  tcsetattr (fd, TCSANOW, &newtio);

  /* Clear the device */
  {
    int flags = fcntl (fd, F_GETFL, 0);
    fcntl (fd, F_SETFL, flags | O_NONBLOCK);
    while (fgets (line, 1024, logfile) );
    fcntl (fd, F_SETFL, flags);
  }

  if (flush)
    setbuf(stdout, NULL);

  while (1)
    {
      FD_ZERO (&rfds);
      FD_SET (fd, &rfds);
      retval = select (fd + 1, &rfds, NULL, NULL, NULL);
      if (retval)
        {
          fgets (line, 1024, logfile);
          fputs (line, stdout);
          if (flush) { fflush(stdout); }
        }
    }

  fclose (logfile);
  tcsetattr (fd, TCSANOW, &oldtio);
  return 0;

}
